contracts/src/v0.8/shared/access/AuthorizedCallers.sol

```diff
@@ -0,0 +1,82 @@
+pragma solidity ^0.8.0;
+
+import {OwnerIsCreator} from "./OwnerIsCreator.sol";
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/structs/EnumerableSet.sol";
+
+/// @title The AuthorizedCallers contract
+/// @notice A contract that manages multiple authorized callers. Enables restricting access to certain functions to a set of addresses.
+contract AuthorizedCallers is OwnerIsCreator {
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  event AuthorizedCallerAdded(address caller);
+  event AuthorizedCallerRemoved(address caller);
+
+  error UnauthorizedCaller(address caller);
+  error ZeroAddressNotAllowed();
+
+  /// @notice Update args for changing the authorized callers
+  struct AuthorizedCallerArgs {
+    address[] addedCallers;
+    address[] removedCallers;
+  }
+
+  /// @dev Set of authorized callers
+  EnumerableSet.AddressSet internal s_authorizedCallers;
+
+  /// @param authorizedCallers the authorized callers to set
+  constructor(address[] memory authorizedCallers) {
+    _applyAuthorizedCallerUpdates(
+      AuthorizedCallerArgs({addedCallers: authorizedCallers, removedCallers: new address[](0)})
+    );
+  }
+
+  /// @return authorizedCallers Returns all authorized callers
+  function getAllAuthorizedCallers() external view returns (address[] memory) {
+    return s_authorizedCallers.values();
+  }
+
+  /// @notice Updates the list of authorized callers
+  /// @param authorizedCallerArgs Callers to add and remove. Removals are performed first.
+  function applyAuthorizedCallerUpdates(AuthorizedCallerArgs memory authorizedCallerArgs) external onlyOwner {
+    _applyAuthorizedCallerUpdates(authorizedCallerArgs);
+  }
+
+  /// @notice Updates the list of authorized callers
+  /// @param authorizedCallerArgs Callers to add and remove. Removals are performed first.
+  function _applyAuthorizedCallerUpdates(AuthorizedCallerArgs memory authorizedCallerArgs) internal {
+    address[] memory removedCallers = authorizedCallerArgs.removedCallers;
+    for (uint256 i = 0; i < removedCallers.length; ++i) {
+      address caller = removedCallers[i];
+
+      if (s_authorizedCallers.remove(caller)) {
+        emit AuthorizedCallerRemoved(caller);
+      }
+    }
+
+    address[] memory addedCallers = authorizedCallerArgs.addedCallers;
+    for (uint256 i = 0; i < addedCallers.length; ++i) {
+      address caller = addedCallers[i];
+
+      if (caller == address(0)) {
+        revert ZeroAddressNotAllowed();
+      }
+
+      s_authorizedCallers.add(caller);
+      emit AuthorizedCallerAdded(caller);
+    }
+  }
+
+  /// @notice Checks the sender and reverts if it is anyone other than a listed authorized caller.
+  function _validateCaller() internal view {
+    if (!s_authorizedCallers.contains(msg.sender)) {
+      revert UnauthorizedCaller(msg.sender);
+    }
+  }
+
+  /// @notice Checks the sender and reverts if it is anyone other than a listed authorized caller.
+  modifier onlyAuthorizedCallers() {
```

Comment:
q: is it common practice to use both a modifier and internal function, versus just using `_validateCaller()` instead of the modifier?

Reply:
In base contracts like this I prefer using a modifier

Reply:
I proposed an internal function because you might have a case where Contract A calls Contract B (e.g. in the case of an OffRamp / OnRamp via the multi-ARL route), and you could have ~50-80% of the scenarios performing no state change at all - in which case you do not need the `isValidatedCaller` check, so you can skip it and save an external contract storage read (I assume the saving would be 5k gas), i.e.:

```solidity

func onIncomingMessage(message) {
  // Skips storage reads
  if (message.destTokenAmounts == 0) return;

  _isCallerAuthorized();
}

```

Reply:
Imo this is still canonical and useful to have both options.